data/src/main/java/me/androidbox/data/worker_manager/SyncFullAgendaItemsWorker.kt

```diff
@@ -0,0 +1,127 @@
+    private val agendaService: AgendaService,
+    private val eventDao: EventDao,
+    private val taskDao: TaskDao,
+    private val reminderDao: ReminderDao,
+) : CoroutineWorker(context, workerParameters) {
+
+    override suspend fun doWork(): Result {
+
+        val responseResult = checkResult {
+            agendaService.fullAgenda()
+        }
+
+        if(runAttemptCount > RETRY_COUNT) {
+            Result.failure()
+        }
+
+        val result = responseResult.fold(
+            onSuccess = { fullAgendaDto ->
+                insertAllAgendaItems(fullAgendaDto, eventDao, taskDao, reminderDao)
+                Result.success()
+            },
+            onFailure = {
+                Result.retry()
+            }
+        )
+
+        return result
+    }
+}
+
+private suspend fun insertAllAgendaItems(fullAgendaDto: FullAgendaDto, eventDao: EventDao, taskDao: TaskDao, reminderDao: ReminderDao) {
+    supervisorScope {
+        val eventJob = launch {
+            fullAgendaDto.events.mapAsync { eventDto ->
+                val eventEntity = EventEntity(
+                    id = eventDto.id,
+                    title = eventDto.title,
+                    description = eventDto.description,
+                    startDateTime = eventDto.from,
+                    endDateTime = eventDto.to,
+                    remindAt = eventDto.remindAt,
+                    eventCreatorId = eventDto.host,
+                    isUserEventCreator = eventDto.isUserEventCreator,
+                    photos = eventDto.photos.map { photoDto ->  photoDto.key },
+                    isGoing = true,
+                    attendees = eventDto.attendees.map { attendeeDto ->
+                        attendeeDto.toAttendee()
+                    }
+                )
+
+                eventDao.insertEvent(eventEntity)
+            }
+        }
+
+        val taskJob = launch {
+            fullAgendaDto.tasks.mapAsync { taskDto ->
+                val taskEntity = TaskEntity(
+                    id = taskDto.id,
+                    title = taskDto.title,
+                    description = taskDto.description,
+                    time = taskDto.time,
+                    remindAt = taskDto.remindAt,
+                    isDone = taskDto.isDone
+                )
+
+                taskDao.insertTask(taskEntity)
+            }
+        }
+
+        val reminderJob = launch {
+            fullAgendaDto.reminders.mapAsync { reminderDto ->
+                val reminderEntity = ReminderEntity(
+                    id = reminderDto.id,
+                    title = reminderDto.title,
+                    description = reminderDto.description,
+                    time = reminderDto.time,
+                    remindAt = reminderDto.remindAt
+                )
+
+                reminderDao.insertReminder(reminderEntity)
+            }
+        }
+
+        eventJob.join()
+        taskJob.join()
+        reminderJob.join()
+    }
+}
```

Comment:
Is this correct for launching multiple coroutines and then joining all of them?

Reply:
Yea it's fine

What I like to do, is:

```
(eventJob + taskJob + reminderJob).forEach { it.join() }
```

Reply:
@MKiperszmid I get the following error
`Using 'plus(Job): Job' is an error. Operator '+' on two Job objects is meaningless. Job is a coroutine context element and `+` is a set-sum operator for coroutine contexts. The job to the right of `+` just replaces the job the left of `+``

Reply:
It doesn't work like @MKiperszmid said. Works only if you're having lists. You'd need to use `listOf(...)` and then you can just call the 3 joins right away

Reply:
100% right. You would need to have it as a list. My bad!